src/test/java/fi/helsinki/cs/tmc/core/commands/DownloadAdaptiveExerciseTest.java

```diff
@@ -0,0 +1,144 @@
+    Course mockCourse;
+    @Mock
+    Exercise mockExerciseOne;
+    @Mock
+    Callable<Exercise> mockGetAdaptiveExercise;
+
+    private Command<Exercise> command;
+    TaskExecutor langs;
+    Path arithFuncsTempDir;
+
+    @Before
+    public void setUp() throws IOException {
+        MockitoAnnotations.initMocks(this);
+        langs = spy(new TaskExecutorImpl());
+        TmcSettingsHolder.set(settings);
+        TmcLangsHolder.set(langs);
+        arithFuncsTempDir = testFolder.getRoot().toPath().resolve("arith_funcs");
+        command = new DownloadAdaptiveExercise(mockObserver, factory);
+
+        doCallRealMethod().when(langs).extractProject(any(Path.class), any(Path.class));
+        mockExerciseOne.setName("ex1");
+        mockExerciseOne.setCourseName("course1");
+
+    }
+
+    @Test
+    public void checkExerciseZipUrl() throws Exception {
+        setUpMocks();
+        Exercise exercise = command.call();
+    }
+
+    @Test
+    public void testDownloadAndExtractSuccess() throws Exception {
+        setUpMocks();
+
+        Exercise exercise = command.call();
+
+        verify(factory).getAdaptiveExercise();
+        verify(factory).getDownloadingExerciseZipTask(mockExerciseOne);
+
+        verifyNoMoreInteractions(factory);
+
+        assertTrue(Files.exists(arithFuncsTempDir));
+    }
+
+    @Test
+    public void testDownloadAndExtractFailure() throws Exception {
+        setUpMocks();
+        when(mockGetAdaptiveExercise.call()).thenReturn(null);
+
+        Exercise exercise = command.call();
+
+        verify(factory).getAdaptiveExercise();
+
+        verifyNoMoreInteractions(factory);
+
+    }
+
+    private void setUpMocks() throws Exception {
+        verifyZeroInteractions(langs);
+
+        when(factory.getAdaptiveExercise()).thenReturn(mockGetAdaptiveExercise);
+        when(mockGetAdaptiveExercise.call()).thenReturn(mockExerciseOne);
+
+        when(mockExerciseOne.getExtractionTarget(any(Path.class))).thenReturn(arithFuncsTempDir);
+        when(settings.getTmcProjectDirectory()).thenReturn(testFolder.getRoot().toPath());
+
+        when(factory.getDownloadingExerciseZipTask(mockExerciseOne))
+                .thenReturn(
+                        new Callable<byte[]>() {
+                            @Override
+                            public byte[] call() throws Exception {
+                                return Files.readAllBytes(
+                                        TestUtils.getZip(this.getClass(), "arith_funcs.zip"));
+                            }
+                        });
+    }
+
+    @Test
+    public void testDownloadAndExtractSuccessWithRealZip() throws Exception {
+        verifyZeroInteractions(langs);
+        TmcServerCommunicationTaskFactory realFactory = new TmcServerCommunicationTaskFactory();
+        assertNotNull(TmcSettingsHolder.get());
+        command = new DownloadAdaptiveExercise(mockObserver, realFactory);
+        Path testPath = Paths.get(System.getProperty("user.dir"));
```

Comment:
Don't add stuff to my home directory and **especially** do not delete stuff from there (Line 142).

Use [`java.nio.file.Files#createTempDirectory`](http://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createTempDirectory-java.lang.String-java.nio.file.attribute.FileAttribute...-) instead, specifically the version that uses the system's default temporary directory.